rfc43: add new RFC for job list service (#408)
flux-framework/rfc: spec_43.rst

- a job submitted without dependencies will never have *dependencies* set
- a job cannot have belong in a *queue* on a system without a job queue
- *exception_type* will only exist if *exception_occurred* is true

If an *attribute* has not been set for a job, it SHALL NOT be returned in the returned data object.

Constraint Operators
====================

Using the constraint syntax described by RFC31, jobs can be filtered
based on the following constraint operators.

 - ``userid``: The set of *integer* values SHALL designate one or more userids
   and match jobs submitted by those userids.

 - ``name``: The set of *string* values SHALL designate one or more job names
   and match jobs with those job names.

 - ``queue``: The set of *string* values SHALL designate one or more queues
   and match jobs with those job queues.

 - ``states``: The set of *integer* values SHALL designate one or more job states
   and match jobs in those job states.

 - ``results``: The set of *integer* values SHALL designate one or more job results

>    and match jobs with those results.

Reviewer: Question: Can the `states` and `results` values be a bitmask of states to match? I.e. is `{"states": [48, 14]}` equivalent to `{"states": [62]}`? If so, it might be useful to call that out specifically here.

Author: yes!  and i forgot that string inputs are ok too (i.e. "depend")

.. github display
   GitHub is NOT the preferred viewer for this file. Please visit
   https://flux-framework.rtfd.io/projects/flux-rfc/en/latest/spec_41.html

43/Job List Service
###################

The Flux Job List Service provides read-only summary information for jobs.

.. list-table::
  :widths: 25 75

  * - **Name**
    - github.com/flux-framework/rfc/spec_43.rst
  * - **Editor**
    - Albert Chu <chu11@llnl.gov>
  * - **State**
    - raw

Language
********

.. include:: common/language.rst

Related Standards
*****************

- :doc:`spec_18`
- :doc:`spec_20`
- :doc:`spec_21`
- :doc:`spec_25`
- :doc:`spec_26`
- :doc:`spec_27`
- :doc:`spec_29`
- :doc:`spec_31`
- :doc:`spec_41`

Background
**********

Users are interested in seeing jobs that have been submitted to the

> job-manager.  Some reasons may include:

Author: Sounds good, although I think your wording suggests this is a complete set of use cases.   I used the phrase "Some use cases include:" instead.

- See which jobs are pending, running, or inactive
- See what jobs are running on specific nodes
- Get general information about a job, such as a job's exit code
- See the order in which jobs were submitted
- See how many jobs are pending in the queue before a specific one

While the job info service described in RFC41 is capable of providing job owners information about their
own jobs, it has several limitations:

- job information may not be easily parsed or collated from multiple sources into one easily parsable format
- information from multiple jobs is not collated into a simple to parse list
- information about the jobs of other users in a multi-user instance is not available

Goals
*****

- Provide read-only access to non-sensitive information for all jobs.

- Hide the complexity of parsing or collating data from multiple sources for commonly accessed information.

- Provide ways to query jobs callers are interested in.

Implementation
**************

> The job list service SHALL provide callers the ability to read job information via identifier keys, which will be called *attributes*.  See `Job Attributes` below for details.

Author: Hmmm I tried the section references earlier but it didn’t work.   Lemme try again.

The job list service SHALL provide a RFC31 constraint syntax for filtering jobs.  See `Constraint Operators` below for details.

Job Attributes
==============

> Job information is defined by the following *attribute* keys listed below.

Author: ehhh, it feels empty without something before the table.  Perhaps a simpler sentence: `The job list service SHALL support the following attributes.`

- a job cannot belong in a *queue* on a system without a job queue
- *exception_type* will only exist if *exception_occurred* is true

If an *attribute* has not been set for a job, it SHALL NOT be returned in the returned data object.

Constraint Operators
====================

Using the constraint syntax described by RFC31, jobs can be filtered
based on the following constraint operators.

``userid``
    Designate one or more userids (*integer*) and match jobs submitted by those userids.

``name``
    Designate one or more job names (*string*) and match jobs with those job names.

``queue``
    Designate one or more queues (*string*) and match jobs submitted to those job queues.

``states``
    Designate one or more job states (*string* or *integer*) and match jobs in those job states.  Both bitmasks (including multiple states) and string names of the states SHALL be accepted.

``results``
    Designate one or more job results (*string* or *integer*) and match jobs with those results.  Both bitmasks (including multiple results) and string names of the results SHALL be accepted.

``t_submit``, ``t_depend``, ``t_run``, ``t_cleanup``, ``t_inactive``

>     Designate one timestamp with a prefixed comparison operator (*string*).  The accepted comparison operators SHALL be `>`, `<`, `>=`, and `<=`, for greater than, less than, greater than or equal, or less than or equal. A timestamp operator SHALL match jobs where the respective timestamp matches against the provided timestamp.

Reviewer: Suggestion: "Designate one timestamp with an _optional_ prefixed comparison operator" ?

Author: actually it's not optional ... I think we decided to go with that to avoid confusion of "what is the default?"

.. github display
   GitHub is NOT the preferred viewer for this file. Please visit
   https://flux-framework.rtfd.io/projects/flux-rfc/en/latest/spec_41.html

43/Job List Service
###################

The Flux Job List Service provides summary information for jobs in the
system.  It provides read-only access.  Several ways to find / filter

> jobs is also supported.

Reviewer: See above comment.

.. list-table::
  :widths: 25 75

  * - **Name**
    - github.com/flux-framework/rfc/spec_43.rst
  * - **Editor**
    - Albert Chu <chu11@llnl.gov>
  * - **State**
    - raw

Language
********

.. include:: common/language.rst

Related Standards
*****************

- :doc:`spec_18`
- :doc:`spec_20`
- :doc:`spec_21`
- :doc:`spec_25`
- :doc:`spec_26`
- :doc:`spec_27`
- :doc:`spec_29`
- :doc:`spec_31`
- :doc:`spec_41`

Background
**********

Users are interested in seeing jobs that have been submitted to the

> scheduler.  Some reason may include:

Reviewer: Minor, but jobs are technically submitted to the job manager not the scheduler (a scheduler need not be loaded to submit and query or list jobs). 

Also "Some reason may include:" :arrow_right: "Some reasons may include"

- See which jobs are pending, running, or inactive
- See what jobs are running on specific nodes
- Get general information about a job, such as a job's exit code
- See the order in which jobs were submitted
- See how many jobs are pending in the queue before a specific one

While the job info service described in RFC41 is capable of providing job owners information about their
own jobs, it has several limitations:

> - job information may not be easily parsed / collated from multiple sources into one easily parsable format

Reviewer: Again, I'd replace `/` with `or` here.

- information from multiple jobs is not collated into a simple to parse list

> - information about non-owned jobs is not available

Reviewer: Suggestion: something like?
```suggestion
- information about the jobs of other users in a multi-user instance is not available
```

Goals
*****

- Provide read-only access to non-sensitive information for all jobs.

- Hide the complexity of parsing or collating data from multiple sources for commonly accessed information.

> - Provide ways to find and/or filter jobs callers are interested in.

Reviewer: The difference between find and filter may not be clear. Maybe replace "find and/or filter" with just "filter" or "query"? I don't have a strong opinion on this one though

.. list-table::
   :header-rows: 1

   * - Attribute
     - Description
     - Value Encoding
   * - id
     - job id
     - integer
   * - userid
     - userid of job submitter
     - integer
   * - urgency
     - job urgency
     - integer
   * - priority
     - job priority
     - integer
   * - t_submit
     - time job was submitted
     - real
   * - t_depend
     - time job entered depend state
     - real
   * - t_run
     - time job entered run state
     - real
   * - t_cleanup
     - time job entered cleanup state
     - real
   * - t_inactive
     - time job entered inactive state
     - real
   * - state
     - current job state
     - integer
   * - name
     - job name
     - string
   * - cwd
     - job current working directory
     - string
   * - queue
     - job queue
     - string
   * - project
     - job project
     - string
   * - bank
     - job bank
     - string
   * - ntasks
     - job task count
     - integer
   * - ncores
     - job core count
     - integer
   * - nnodes
     - job node count
     - integer
   * - ranks
     - ranks a job ran on
     - integer
   * - nodelist

>      - nodes a job ran on, may accept RFC29 Hostlist

Reviewer: ```suggestion
     - nodes assigned to a job in RFC29 Hostlist format
```
Unless I'm misunderstanding what is meant by "may accept" here (maybe this was meant to describe the constraint not the attribute?)

Author: ahhh yeah I messed up meaning the constraint input

- a job cannot belong in a *queue* on a system without a job queue
- *exception_type* will only exist if *exception_occurred* is true

If an *attribute* has not been set for a job, it SHALL NOT be returned in the returned data object.

Constraint Operators
====================

Using the constraint syntax described by RFC31, jobs can be filtered
based on the following constraint operators.

``userid``
    Designate one or more userids (*integer*) and match jobs submitted by those userids.

``name``
    Designate one or more job names (*string*) and match jobs with those job names.

``queue``
    Designate one or more queues (*string*) and match jobs submitted to those job queues.

``states``
    Designate one or more job states (*string* or *integer*) and match jobs in those job states.  Both bitmasks (including multiple states) and string names of the states SHALL be accepted.

``results``
    Designate one or more job results (*string* or *integer*) and match jobs with those results.  Both bitmasks (including multiple results) and string names of the results SHALL be accepted.

``t_submit``, ``t_depend``, ``t_run``, ``t_cleanup``, ``t_inactive``

>     Designate one timestamp with a prefixed comparison operator (*string*).  The accepted comparison operators SHALL be `>`, `<`, `>=`, and `<=`, for greater than, less than, greater than or equal, or less than or equal.  Match jobs where the respective timestamp matches against the submitted timestamp and comparison operator.

Reviewer: Suggestion: 
> Match jobs where the respective timestamp matches against the submitted timestamp and comparison operator.

Maybe reword with RFC syntax ("A timestamp operator SHALL match...") and replace "submitted timestamp" (could be confused with t_submit or job submission) with "provided timestamp"?
